src/ordered_set_lang.ml

```diff
@@ -297,48 +336,46 @@ module Unexpanded = struct
     in
     loop t.ast Pos init
 
-  let expand t ~dir ~files_contents ~(f : String_with_vars.t -> Value.t list) =
-    let context = t.context in
-    let f_elems s =
-      let loc = String_with_vars.loc s in
-      Ast.union
-        (List.map (f s) ~f:(fun s -> Ast.Element (loc, Value.to_string ~dir s)))
-    in
-    let rec expand (t : ast) : ast_expanded =
+  type expander =
+    { f: 'a. mode:'a String_with_vars.Mode.t
+        -> String_with_vars.t
+        -> ('a, String_with_vars.t) String_with_vars.Partial.t
+    }
+
+  let expand t ~dir ~f =
+    let files = ref Path.Set.empty in
+    let rec expand (t : ast) : Partial.ast =
       let open Ast in
+      let open String_with_vars.Partial in
       match t with
-      | Element s -> f_elems s
+      | Element s ->
+        Element (
+          match f.f ~mode:Many s with
+          | Expanded v -> Expanded (String_with_vars.loc s, v)
+          | Unexpanded _ as v -> v)
       | Standard -> Standard
-      | Include fn ->
-        let sexp =
-          let path =
-            match f fn with
-            | [x] -> Value.to_path ~dir x
-            | _ ->
-              Loc.fail (String_with_vars.loc fn)
-                "An unquoted templated expanded to more than one value. \
-                 A file path is expected in this position."
-          in
-          match Path.Map.find files_contents path with
-          | Some x -> x
-          | None ->
-            Exn.code_error
-              "Ordered_set_lang.Unexpanded.expand"
-              [ "included-file", Path.sexp_of_t path
-              ; "files", Sexp.To_sexp.(list Path.sexp_of_t)
-                           (Path.Map.keys files_contents)
-              ]
-        in
-        let open Stanza.Of_sexp in
-        parse
-          (Parse.without_include ~elt:(String_with_vars.t >>| f_elems))
-          context
-          sexp
+      | Include (Template_path fn) ->
+        begin match f.f ~mode:Single fn with
+        | Unexpanded e ->
+          Loc.fail (String_with_vars.loc fn)
+            "Failed to fully expand include path. Unexpanded value: %a"
```

Comment:
I'm not sure this is very clear for users. It's more of an implementation detail. What about pointing to the first percent form in `e` and saying: `This percent form is not supported in this position`?